src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresDistributedLockDAO.scala

```diff
@@ -0,0 +1,128 @@
+package org.broadinstitute.dsde.workbench.sam.dataAccess
+
+import cats.data.OptionT
+import cats.effect.{Async, Resource}
+import cats.syntax.all._
+import fs2.Stream
+import org.broadinstitute.dsde.workbench.model._
+import org.broadinstitute.dsde.workbench.sam.db.SamParameterBinderFactory._
+import org.broadinstitute.dsde.workbench.sam.db._
+import org.broadinstitute.dsde.workbench.sam.db.tables._
+import org.broadinstitute.dsde.workbench.sam.util.lock.DistributedLockAlgebra
+import org.broadinstitute.dsde.workbench.sam.util.DatabaseSupport
+
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+import scala.concurrent.duration.FiniteDuration
+
+class PostgresDistributedLockDAO[F[_]](
```

Comment:
I toyed with the idea of completely re-writing the logic, but decided against it. This is already a pretty large change in terms of how Sam works, so I wanted to keep the complexity of the change lower. With the DAO implemented, it wouldn't be hard to go in and refactor the actual lock functionality, but I see that as more of a follow-on task.

Essentially, the `DistributedLock` class presented a CATS-ified API, and I didn't want to change both that API and its backend in one swell foop. It would have required a complete rewrite of the tests and a re-wiring of the things that call the lock on top of getting it to use Postgres instead of Firestore. As it stands, this PR does what it sets out to do, which addresses the problems that have been reported by replacing Firestore with Postgres. I was nervous to re-engineer the logic, especially since this started as a side-project and got promoted to a "thing we need to do".